### PR TITLE
Cotel Day 3

### PR DESCRIPTION
**Pull request by Cotel**
# Cotel's solution for day 3 challenge

## Description

This is just Part 1 of the problem.

Given a spiral matrix we have to find the Manhattan Distance for a number included in the matrix to the center:

```
17  16  15  14  13
18   5   4   3  12
19   6   1   2  11
20   7   8   9  10
21  22  23---> ...
```

For example: Distance for 12 is 3

## Solution

Well, I tried to make this functional like but I couldn't imagine how. So the big part for this solution, which is the spiral matrix generation, is horrible 😢 

Anyway, I will only explain the subproblem which I find more interesting:

We need to know how big is the matrix to build it. For example, we know that for including number `12` in the matrix, we need to build a `5x5` square. 

We can know for a matrix of size `n` which will be the number in its inferior right corner as they follow a pattern: 1, 9, 25, 49, ... This is `(2*n + 1)^2`.

So, for a number `x` I generate this pattern until `x <= (2*n + 1)^2`:

```scala
private def calculateArraySize(i: Int, acc: Int = 0): Int = Math.pow(2 * acc + 1, 2).toInt match {
  case corner if corner >= i => Math.sqrt(corner).toInt
  case corner if corner < i => calculateArraySize(i, acc + 1)
}
```

## Tests battery

I've used the provided tests 🤷‍♂️ 

Happy Advent of Code!

### CODE REVIEW

**Comment by Cotel** [comment]
Solution for part 2 is as horredous as the first or even worse so there is no need for commenting it 🤣

File: cotel/src/main/scala/day3/SpiralMatrix.scala [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+  type Matrix[A] = Array[Array[A]]
+
+  def Matrix(xs: Array[Int]*) = Array(xs: _*)
+
+  def distanceToCenterOfMatrix(number: Int): Int = {
+    val spiralMatrix = generateSpiralMatrix(number)
+
+    val numberPosition = get2DPosition(number, spiralMatrix).get
+
+    val middleRow = spiralMatrix.length / 2
```
Comment (tonilopezmr) [inline cotel/src/main/scala/day3/SpiralMatrix.scala:14]: double or int return?

File: cotel/src/main/scala/day3/SpiralMatrix.scala [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        }
+      }
+
+      expo += 2
+    }
+
+    -1
+  }
+
+  private def sumNeighbors(matrix: Matrix[Int], position: (Int, Int)): Int =
```
Comment (tonilopezmr) [inline cotel/src/main/scala/day3/SpiralMatrix.scala:143]: How do you iterate over the matrix?

for i and j it's like a for inside for?

**Review by tonilopezmr — APPROVED** [review]
Fuck third day xD

File: cotel/src/main/scala/day3/SpiralMatrix.scala (hunk 1 above)
Comment (Cotel) [reply cotel/src/main/scala/day3/SpiralMatrix.scala:14]: Both numbers in the division are Int so the result is Int too

File: cotel/src/main/scala/day3/SpiralMatrix.scala (hunk 2 above)
Comment (Cotel) [reply cotel/src/main/scala/day3/SpiralMatrix.scala:143]: Yes

**Pull request merged**